README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# covidphdata: An Interface to the Philippines Department of Health COVID-19 Data Drop <img src="man/figures/logo.png" width="200px" align="right" />

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![R build
status](https://github.com/como-ph/covidphdata/workflows/R-CMD-check/badge.svg)](https://github.com/como-ph/covidphdata/actions)
[![R build
status](https://github.com/como-ph/covidphdata/workflows/test-coverage/badge.svg)](https://github.com/como-ph/covidphdata/actions)
[![Codecov test
coverage](https://codecov.io/gh/como-ph/covidphdata/branch/master/graph/badge.svg)](https://codecov.io/gh/como-ph/covidphdata?branch=master)
[![CodeFactor](https://www.codefactor.io/repository/github/como-ph/covidphdata/badge)](https://www.codefactor.io/repository/github/como-ph/covidphdata)
[![DOI](https://zenodo.org/badge/282435192.svg)](https://zenodo.org/badge/latestdoi/282435192)
<!-- badges: end -->

The [Philippines Department of Health](https://www.doh.gov.ph) has made
COVID-19 related data publicly available as part of its mandate to
promoting transparency and accountability in governance. The Philippines
COVID-19 Data Drop is distributed via *Google Drive* with latest updated
data provided daily (generally). Data from previous days are archived
and also made available through *Google Drive*. This package provides a
coherent, robust and performant API to the latest and archived
Philippines COVID-19 data.

## Motivation and development history

In early April 2020 as part of [CoMo
Philippines](https://como-ph.github.io)’s contribution to the [CoMo
Consortium](https://www.tropicalmedicine.ox.ac.uk/news/como-consortium-the-covid-19-pandemic-modelling-in-context)’s
work in COVID-19 modelling for the Philippines context, we developed an
[R](https://cran.r-project.org) package called
[`comoparams`](https://como-ph.github.io/comoparams) that provided
[R](https://cran.r-project.org) functions for access, handling and
processing of relevant data required for COVID-19 modelling parameters
for the Philippines. Within the
[`comoparams`](https://como-ph.github.io/comoparams) package was a set
of functions developed to interface with the then [newly-announced
Philippines Department of Health Data Drop for
COVID-19](https://www.facebook.com/OfficialDOHgov/posts/announcement-covid-19-data-drop-now-accessible-onlinecovid-19-related-data-now-a/3227811310563432/).
From then until late June 2020, we continued maintenance of the
[`comoparams`](https://como-ph.github.io/comoparams) package.
Specifically, the set of functions for interfacing with the **DoH Data
Drop** required several functions updates (see history of issues and
changes to the package
[here](https://github.com/como-ph/comoparams/issues?q=is%3Aissue)) in
relation to noted issues with the **DoH Data Drop** system and the data
it provides. A good discussion of some of these issues by the **UP
COVID-19 Pandemic Response Team** can be found
[here](https://www.up.edu.ph/prevailing-data-issues-in-the-time-of-covid-19-and-the-need-for-open-data/).
Other issues that required syntax-breaking changes to the
[`comoparams`](https://como-ph.github.io/comoparams) package functions
were related to how the DoH Data Drop was released and distributed.
Earlier during its first release, the **DoH Data Drop** was distributed
through *Google Sheets* via a single non-changing link. This was for
both latest release and the archived data. A few weeks later, the **DoH
Data Drop** was moved to *Google Drive* again via a single non-changing
link for both the latest and the archive data. A month or so after,
release of the latest **DoH Data Drop** was distributed through a single
link that changed during every release (usually daily). The archive
data, on the other hand, was distributed through a different but
constant [link](https://bit.ly/DataDropArchives). The **DoH Data Drop**
distribution system has since remained as such up to date though the
archive data now only includes the current month archive compared to
what it was in late June 2020 when all previous months of archive data
were available.

Currently, given the still persistent issues raised with regard to the
**DoH Data Drop** system and the datasets it distributes, we felt it was
important to create a separate focused [R](https://cran.r-project.org)
package just for accessing, handling and processing of the **DoH Data
Drop** that can be universally used by other
[R](https://cran.r-project.org) users regardless of their ultimate
end-use for the data (e.g., reporting, visualisation, modelling, etc.).
We also wanted to contribute to the work of fellow scientists and
researchers in the Philippines whom we are familiar with and whom we
know use [R](https://cran.r-project.org) for the epidemiologic analysis
work they perform which they share with the general public in the
Philippines. From our own experiences of dealing with the **DoH Data
Drop**, we felt that it would be extremely helpful for fellow scientists
and researchers who use [R](https://cran.r-project.org) to have
consistent and performant data access, handling and processing functions
to interface with the **DoH Data Drop** that can potentially reduce the
daily and/or weekly workload of producing COVID-19 analysis and reports
and streamline their routine analytical workflows.

To this end, we took inspiration from the functions we developed in the
[`comoparams`](https://como-ph.github.io/comoparams) package and
developed this `covidphdata` package based on the following
[R](https://cran.r-project.org) package design principles:

  - use of a **modular and refactorable approach** to determining and
    developing the functions required;

  - creation of an [R](https://cran.r-project.org)-based **robust and
    performant application programme interface or API** to the *Google
    Drive*-based **DoH Data Drop**;

  - application of **modern data handling and processing techniques**
    available in [R](https://cran.r-project.org); and,

  - output of **coherent and compliant data structures** ready for
    various applications.

Finally, we want to contribute to both the *public discourse* and to the
*practice* of **open data** and **open science** in the Philippines and
encourage others to do the same. Our group’s codebase for our work on
COVID-19 modelling (primarily in [R](https://cran.r-project.org)
language for statistical computing) is available publicly via
[GitHub](https://github.com/como-ph) including the one for the
`covidphdata` package (see <https://github.com/como-ph/covidphdata>).
Our hope is that by creating this package and making its use available
to all [R](https://cran.r-project.org) users and its codebase open
source, we can standardise accessing, handling and processing of the
**DoH Data Drop** thereby providing transparency to the approaches
applied to the data before it is analysed or visualised.

## What does `covidphdata`do?

The `covidphdata` primarily provides functions that serve as low-level
wrappers to specific [`googledrive`](http://googledrive.tidyverse.org)
package functions that support the access to files and folders contained
in a *Google Drive* given that the **DoH Data Drop** is distributed
through this platform.

Currently, the `covidphdata` package:

  - Provides functions (`datadrop_id*`) to dynamically retrieve unique
    **DoH Data Drop** file and folder identifiers used by *Google
    Drive*;

  - Provides a function (`datadrop_ls`) to list the files and folders
    within a specified **DoH Data Drop** folder in *Google Drive*;

  - Provides a function (`datadrop_download`) to download a specified
    file within **DoH Data Drop** in *Google Drive*; and,

  - Provides functions (`datadrop_get`) to retrieve a specified file
    within **DoH Data Drop** in *Google Drive* into
    [R](https://cran.r-project.org).

## Installation

<!---
You can install the released version of covidphdata from [CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("covidphdata")
```
--->

`covidphdata` is not yet available on
[CRAN](https://cran.r-project.org).

The development version of `covidphdata` can be installed via GitHub
using the `remotes` package:

``` r
if(!require(remotes)) install.packages("remotes")
remotes::install_github("como-ph/covidphdata")
```

## Usage

### Google authentication

The **DoH Data Drop** is distributed via *Google Drive* as shared
folders and files that can be accessed by anyone with a link. From a
user who accesses the **DoH Data Drop** interactively (i.e., clicking
the **DoH Data Drop** daily link and then accessing the folders and
files on *Google Drive* on the browser), this would mean that one
doesn’t have to be an authorised user to access the files. On the
other hand, for an [R](https://cran.r-project.org) user, it would be
convenient to be able to access and retrieve the **DoH Data Drop**
folders and files within [R](https://cran.r-project.org) without having
to do separate steps of dealing with going to a browser and downloading
the data from the **DoH Data Drop** before being able to access them in
[R](https://cran.r-project.org). This is the main use case for the
`covidphdata` package.

The `covidphdata` is built on the key functionalities of the
`googledrive` package and follows the basic workflow of data retrieval
from *Google Drive*. The `googledrive` package requires authenticating
with *Google Drive* and has functions that facilitate this specifically
its `drive_auth` function. So, the easiest way to authenticate as a
*Google Drive* user is to make a call to the `drive_auth` function as
follows:

``` r
## Load googledrive package
library(googledrive)

## Authenticate with Google
drive_auth()
```

This function when called will open the browser to allow the user to
login their username and password to their *Google* account. This is the
easiest approach to authenticate as most users will be able to implement
this easily. Also, this step will only have to be done per session which
will be enough to use `covidphdata` to retrieve the data one needs into
[R](https://cran.r-project.org). After that, all operations one needs to
do will be with dealing with the data that is within
[R](https://cran.r-project.org) which doesn’t require accessing *Google
Drive* again. However, it is possible that if at some point later within
the same session you need to access different dataset from the archive
that you will need to authenticate again. If this is the case, you will
be asked on your [R](https://cran.r-projectd.org) console to confirm the
email account that you previously logged in with and your
[R](https://cran.r-project.org) session will continue to use those
authentication details.

This is the most straightforward and easiest way to authenticate with
*Google Drive* for most users. However, if you prefer a non-interactive
way of authenticating, read this [good
article](https://googledrive.tidyverse.org/articles/articles/bring-your-own-app.html)
from the `googledrive` package creators on the alternative ways to
authenticate that doesn’t require interactive input.

Once authenticated, one can now proceed with the `covidphdata` data
retrieval workflow outlined below.

### The `covidphdata` data retrieval workflow

The different functions currently available through `covidphdata` are
linked through the following general workflow:

``` r
## Load covidphdata and googledrive package
library(covidphdata)
library(googledrive)

## Step 1: Get Google Drive ID for latest DoH Data Drop
gid <- datadrop_id()

## Step 2: List the files and folders available in the latest DoH Data Drop
data_list <- datadrop_ls(id = gid)

## Step 3: Retrieve the specified/required dataset and load into R
datadrop_get(tbl = data_list, fn = "Case Information", path = tempfile())
```

This workflow produces the following output:

    #> # A tibble: 416,852 x 22
    #>    CaseCode   Age AgeGroup Sex   DateSpecimen DateResultRelea… DateRepConf
    #>    <chr>    <dbl> <chr>    <chr> <chr>        <chr>            <chr>      
    #>  1 C581144     53 50 to 54 MALE  "2020-07-15" "2020-07-16"     2020-07-21 
    #>  2 C352619     56 55 to 59 FEMA… ""           ""               2020-05-04 
    #>  3 C466590     36 35 to 39 MALE  "2020-07-14" "2020-07-17"     2020-07-20 
    #>  4 C971812     44 40 to 44 MALE  "2020-05-07" ""               2020-05-21 
    #>  5 C445586     57 55 to 59 MALE  "2020-07-14" "2020-07-16"     2020-07-19 
    #>  6 C354376     34 30 to 34 FEMA… "2020-04-14" "2020-04-16"     2020-09-01 
    #>  7 C583142     54 50 to 54 MALE  "2020-07-20" "2020-07-25"     2020-08-02 
    #>  8 C329630     56 55 to 59 FEMA… "2020-07-20" "2020-07-22"     2020-07-27 
    #>  9 C349878     22 20 to 24 MALE  "2020-07-22" "2020-07-26"     2020-07-30 
    #> 10 C956795     48 45 to 49 FEMA… "2020-07-30" "2020-08-01"     2020-08-04 
    #> # … with 416,842 more rows, and 15 more variables: DateDied <chr>,
    #> #   DateRecover <chr>, RemovalType <chr>, Admitted <chr>, RegionRes <chr>,
    #> #   ProvRes <chr>, CityMunRes <chr>, CityMuniPSGC <chr>, BarangayRes <chr>,
    #> #   BarangayPSGC <chr>, HealthStatus <chr>, Quarantined <chr>, DateOnset <chr>,
    #> #   Pregnanttab <chr>, ValidationStatus <chr>

The functions in the `covidphdata` workflow were designed to be usable
with piped operations provided by the
[`magrittr`](https://magrittr.tidyverse.org) package. The pipe operator
`%>%` can be used with the `covidphdata` workflow as follows:

``` r
## Load magrittr package
library(magrittr)

## Retrieve latest Case Information dataset from DoH Data Drop
datadrop_id() %>%                                 ## Step 1
  datadrop_ls() %>%                               ## Step 2
  datadrop_get(fn = "Case Information",           ## Step 3
               path = tempfile())                 
```

This outputs this result:

    #> # A tibble: 416,852 x 22
    #>    CaseCode   Age AgeGroup Sex   DateSpecimen DateResultRelea… DateRepConf
    #>    <chr>    <dbl> <chr>    <chr> <chr>        <chr>            <chr>      
    #>  1 C581144     53 50 to 54 MALE  "2020-07-15" "2020-07-16"     2020-07-21 
    #>  2 C352619     56 55 to 59 FEMA… ""           ""               2020-05-04 
    #>  3 C466590     36 35 to 39 MALE  "2020-07-14" "2020-07-17"     2020-07-20 
    #>  4 C971812     44 40 to 44 MALE  "2020-05-07" ""               2020-05-21 
    #>  5 C445586     57 55 to 59 MALE  "2020-07-14" "2020-07-16"     2020-07-19 
    #>  6 C354376     34 30 to 34 FEMA… "2020-04-14" "2020-04-16"     2020-09-01 
    #>  7 C583142     54 50 to 54 MALE  "2020-07-20" "2020-07-25"     2020-08-02 
    #>  8 C329630     56 55 to 59 FEMA… "2020-07-20" "2020-07-22"     2020-07-27 
    #>  9 C349878     22 20 to 24 MALE  "2020-07-22" "2020-07-26"     2020-07-30 
    #> 10 C956795     48 45 to 49 FEMA… "2020-07-30" "2020-08-01"     2020-08-04 
    #> # … with 416,842 more rows, and 15 more variables: DateDied <chr>,
    #> #   DateRecover <chr>, RemovalType <chr>, Admitted <chr>, RegionRes <chr>,
    #> #   ProvRes <chr>, CityMunRes <chr>, CityMuniPSGC <chr>, BarangayRes <chr>,
    #> #   BarangayPSGC <chr>, HealthStatus <chr>, Quarantined <chr>, DateOnset <chr>,
    #> #   Pregnanttab <chr>, ValidationStatus <chr>

The resulting dataset retrieved by the piped operation is exactly the
same as the first example. However, the piped operation workflow is more
streamlined. Either of these approaches can be used when retrieving
datasets from the **DoH Data Drop**.

#### Step 1: Get Google Drive ID for DoH Data Drop

The Philippines [Department of Health (DoH)](https://www.doh.gov.ph)
currently distributes the latest **Data Drop** via a fixed shortened URL
([bit.ly/DataDropPH](https://bit.ly/DataDropPH)) which links/points to a
new *Google Drive* endpoint daily or whenever the daily updated **DoH
Data Drop** is available. This *Google Drive* endpoint is a **README**
document in portable document format (PDF) which contains a *privacy and
confidentiality statement*, *technical notes with regard to the latest
data*, *technical notes with regard to previous (archive data)* and *two
shortened URLs* - one linking to the *Google Drive* folder that contains
all the latest officially released datasets, and the other linking to
the datasets released previously (archives). Of these, the first
shortened URL linking to the *Google Drive* folder containing the latest
officially released datasets is different for every release and can only
be obtained through the **README** document released for a specific day.

The function `datadrop_id_latest()` reads that PDF file, extracts the
shortened URL for the latest official released datasets written in that
file, expands that shortened URL and then extracts the unique *Google
Drive* ID for the latest officially released datasets. With this *Google
Drive* ID, other functions can then be used to retrieve information and
data from the *Google Drive* specified by this ID.

The **DoH Data Drop** archives, on the other hand, is distributed via a
fixed shortened URL ([bit.ly/DataDropArchives](bit.ly/DataDropArchives))
which links/points to a *Google Drive* folder containing the previous
**DoH Data Drop** releases.

The function `datadrop_id_archive` expands that shortened URL and then
extracts the unique *Google Drive* ID for the **DoH Data Drop** archives
folder. With this *Google Drive* ID, other functions can then be used to
retrieve information and data from the *Google Drive* specified by this
ID.

The function `datadrop_id` wraps these two `datadrop_id_*` functions and
outputs the appropriate *Google Drive* ID based on specified parameters.
The default output of `datadrop_id` is the *Google Drive* ID of the
latest officially released datasets as shown below:

``` r
datadrop_id()
#> [1] "1N_u-9xjcatJJ08JqUKwpq2eFku_LWOjl"
```

To get the *Google Drive* ID of a **DoH Data Drop** archive, the
following parameters need to be set:

``` r
datadrop_id(version = "archive", .date = "2020-11-01")
#> [1] "1O2Gt_MUPKtWWPK6ainagRiehK0pCH7Nv"
```

#### Step 2: List the files and folders available in the latest DoH Data Drop

Once an appropriate *Google Drive* ID is obtained, this can be used with
the `datadrop_ls` function to retrieve a listing of all the files and
folders within the specified *Google Drive*:

``` r
## List the files and folders inside the latest Google Drive DoH Data Drop
gid <- datadrop_id()
datadrop_ls(id = gid)
#> # A tibble: 15 x 3
#>    name                                      id                  drive_resource 
#>  * <chr>                                     <chr>               <list>         
#>  1 DOH COVID Data Drop_ 20201121 - 09 Quara… 1hWAgqykG3kPKT5Fgr… <named list [3…
#>  2 DOH COVID Data Drop_ 20201121 - 04 Case … 1nyEEw4hrwHuq3Oc1y… <named list [3…
#>  3 DOH COVID Data Drop_ 20201121 - 02 Metad… 1Umc2_nxW2I1ZLjmUA… <named list [3…
#>  4 DOH COVID Data Drop_ 20201121 - 08 Quara… 1RGyLcAevZZ2YITRS7… <named list [3…
#>  5 DOH COVID Data Drop_ 20201121 - 10 DOH D… 16byfiN1KVYENxhW39… <named list [3…
#>  6 DOH Data Drop 20201121 - Changelog.xlsx   1TAVweGNepRkvGOtIX… <named list [3…
#>  7 DOH COVID Data Drop_ 20201121 - 06 DOH D… 1Epa-Bz4WdjXK9rLH-… <named list [3…
#>  8 DOH COVID Data Drop_ 20201121 - 03 Metad… 1_WxunmnV9PLAHzVtZ… <named list [3…
#>  9 DOH COVID Data Drop_ 20201121 - 05 DOH D… 1AA6keECjIwUQkzVet… <named list [3…
#> 10 DOH COVID Data Drop_ 20201121 - 07 Testi… 1yJqWFOpB-X7gBkzSH… <named list [3…
#> 11 DOH COVID Data Drop_ 20201121 - 12 DDC T… 1pMrknQpIR7XJuLeFA… <named list [3…
#> 12 DOH COVID Data Drop_ 20201121 - 02 Metad… 1CDs8lJ-aM-MnWQOlR… <named list [3…
#> 13 DOH COVID Data Drop_ 20201121 - 03 Metad… 1rlhqnLZyU3Ri_COgC… <named list [3…
#> 14 DOH COVID Data Drop_ 20201121 - 11 DOH D… 1OtL3G_bN2Ri1C6-WZ… <named list [3…
#> 15 01 READ ME FIRST (11_21).pdf              1ywcS448kmrWZra2vu… <named list [3…
```

``` r
## List the files and folders inside the Google Drive DoH Data Drop on 
## 1 November 2020
gid <- datadrop_id(version = "archive", .date = "2020-11-01")
datadrop_ls(id = gid)
#> # A tibble: 13 x 3
#>    name                                      id                  drive_resource 
#>  * <chr>                                     <chr>               <list>         
#>  1 DOH COVID Data Drop_ 20201101 - 02 Metad… 1m8zDRFJ5qKHky60zb… <named list [3…
#>  2 DOH Data Drop 20201101 - Changelog.xlsx   1KHWj0Mo__y0UqkXfm… <named list [3…
#>  3 01 READ ME FIRST (11_01).pdf              1sipJeeWgMqmS5cYv5… <named list [3…
#>  4 DOH COVID Data Drop_ 20201101 - 12 DDC T… 1DRYzj6m7HALBM2sVy… <named list [3…
#>  5 DOH COVID Data Drop_ 20201101 - 06 DOH D… 1XAo_lunz5HXZpGnF1… <named list [3…
#>  6 DOH COVID Data Drop_ 20201101 - 09 Quara… 1YarEezOL_wtdtBhFZ… <named list [3…
#>  7 DOH COVID Data Drop_ 20201101 - 11 DOH D… 1_Y8dx3ogvJN0RMHYv… <named list [3…
#>  8 DOH COVID Data Drop_ 20201101 - 04 Case … 1G3RITnBxmO0qIYAOG… <named list [3…
#>  9 DOH COVID Data Drop_ 20201101 - 10 DOH D… 16KEnIvD4ol4oQrtu8… <named list [3…
#> 10 DOH COVID Data Drop_ 20201101 - 08 Quara… 19qpm1UseQrSmELji8… <named list [3…
#> 11 DOH COVID Data Drop_ 20201101 - 03 Metad… 1a4yFn9jxGgDQ3_StU… <named list [3…
#> 12 DOH COVID Data Drop_ 20201101 - 05 DOH D… 1D6sO0tNudLENeiMzK… <named list [3…
#> 13 DOH COVID Data Drop_ 20201101 - 07 Testi… 1eW8C062WTfMJC9mn2… <named list [3…
```

#### Step 3: Retrieve the specified/required dataset and load into R

Finally, using the specific *Google Drive* ID for the file of interest,
the `datadrop_id_file` function is used to retrieve the file and output
it into [R](https://cran.r-project.org):

``` r
## Retrieve the latest Case Information file in the DoH Data Drop
gid <- datadrop_id()
tab <- datadrop_ls(id = gid)
datadrop_get(tbl = tab, fn = "Case Information", path = tempfile())
#> # A tibble: 416,852 x 22
#>    CaseCode   Age AgeGroup Sex   DateSpecimen DateResultRelea… DateRepConf
#>    <chr>    <dbl> <chr>    <chr> <chr>        <chr>            <chr>      
#>  1 C581144     53 50 to 54 MALE  "2020-07-15" "2020-07-16"     2020-07-21 
#>  2 C352619     56 55 to 59 FEMA… ""           ""               2020-05-04 
#>  3 C466590     36 35 to 39 MALE  "2020-07-14" "2020-07-17"     2020-07-20 
#>  4 C971812     44 40 to 44 MALE  "2020-05-07" ""               2020-05-21 
#>  5 C445586     57 55 to 59 MALE  "2020-07-14" "2020-07-16"     2020-07-19 
#>  6 C354376     34 30 to 34 FEMA… "2020-04-14" "2020-04-16"     2020-09-01 
#>  7 C583142     54 50 to 54 MALE  "2020-07-20" "2020-07-25"     2020-08-02 
#>  8 C329630     56 55 to 59 FEMA… "2020-07-20" "2020-07-22"     2020-07-27 
#>  9 C349878     22 20 to 24 MALE  "2020-07-22" "2020-07-26"     2020-07-30 
#> 10 C956795     48 45 to 49 FEMA… "2020-07-30" "2020-08-01"     2020-08-04 
#> # … with 416,842 more rows, and 15 more variables: DateDied <chr>,
#> #   DateRecover <chr>, RemovalType <chr>, Admitted <chr>, RegionRes <chr>,
#> #   ProvRes <chr>, CityMunRes <chr>, CityMuniPSGC <chr>, BarangayRes <chr>,
#> #   BarangayPSGC <chr>, HealthStatus <chr>, Quarantined <chr>, DateOnset <chr>,
#> #   Pregnanttab <chr>, ValidationStatus <chr>
```

``` r
## Retrieve the archive Case Information file in the DoH Data Drop
## on 1 November 2020
gid <- datadrop_id(version = "archive", .date = "2020-11-01")
tab <- datadrop_ls(id = gid)
datadrop_get(tbl = tab, fn = "Case Information", path = tempfile())
#> # A tibble: 383,113 x 22
#>    CaseCode   Age AgeGroup Sex   DateSpecimen DateResultRelea… DateRepConf

NEWS.md
# covidphdata 0.1.0

Pre-release of `covidphdata`.

## Features

* Provides functions (`datadrop_id*`) to dynamically retrieve unique **DoH Data Drop** file and folder identifiers used by *Google Drive*;

* Provides a function (`datadrop_ls`) to list the files and folders within a specified **DoH Data Drop** folder in *Google Drive*;

* Provides a function (`datadrop_download`) to download a specified file within **DoH Data Drop** in *Google Drive*; and,

* Provides functions (`datadrop_get`) to retrieve a specified file within **DoH Data Drop** in *Google Drive* into [R](https://cran.r-project.org).


R/get_id.R
################################################################################
#
#'
#' Get Google Drive ID for latest or archive DoH Data Drop folders
#'
#' The **DoH Data Drop** is distributed using *Google Drive* with the latest
#' data released through a new *Google Drive* folder and the older data archived
#' into the same persistent *Google Drive* folder.
#'
#' The Philippines Department of Health (DoH) currently distributes the latest
#' **Data Drop** via a fixed shortened URL ([bit.ly/DataDropPH](https://bit.ly/DataDropPH))
#' which links/points to a new *Google Drive* endpoint daily or whenever the
#' daily updated data drop is available. This *Google Drive* endpoint is a
#' **README** document in portable document format (PDF) which contains a
#' privacy and confidentiality statement, technical notes with regard to the
#' latest data, technical notes with regard to previous (archive data) and two
#' shortened URLs - one linking to the *Google Drive* folder that contains all
#' the latest officially released datasets, and the other linking to the
#' datasets released previously (archives). Of these, the first shortened URL
#' linking to the *Google Drive* folder containing the latest officially released
#' datasets is different for every release and can only be obtained through the
#' **README** document released for a specific day.
#'
#' The function [datadrop_id_latest()] reads the **README** PDF file, extracts the
#' shortened URL for the latest official released datasets written in that file,
#' expands that shortened URL and then extracts the unique *Google Drive* ID for
#' the latest officially released datasets. With this *Google Drive* ID, other
#' functions can then be used to retrieve information and data from the Google
#' Drive specified by this ID.
#'
#' The **DoH Data Drop** archives, on the other hand, is distributed via a fixed
#' shortened URL ([bit.ly/DataDropArchives](https://bit.ly/DataDropArchives))
#' which links/points to a *Google Drive* folder containing the previous
#' **DoH Data Drop** releases.
#'
#' The function [datadrop_id_archive()] expands that shortened URL and then
#' extracts the unique *Google Drive* ID for the **DoH Data Drop** archives folder.
#' With this *Google Drive* ID, other functions can then be used to retrieve
#' information and data from the *Google Drive* specified by this ID.
#'
#' @param verbose Logical. Should message on operation progress be shown.
#'   Default is TRUE.
#' @param version A character value specifying whether to get the latest
#'   available **DoH Data Drop** (`latest`) or to get **DoH Data Drop** archive
#'   (`archive`). Default to `latest`.
#' @param .date A character value for date in *YYYY-MM-DD* format. This is the
#'   date for the archive **DoH Data Drop** for which an ID is to be returned.
#'   Should be specified when using [datadrop_id_archive()]. For [datadrop_id()],
#'   only used when `version` is set to `archive` otherwise ignored.
#'
#' @return A 33-character string for the *Google Drive* ID of the latest
#'   **DoH Data Drop** or the archive **DoH Data Drop**
#'
#' @author Ernest Guevarra
#'
#' @examples
#' \dontrun{
#'   library(googledrive)
#'
#'   ## Deauthorise
#'   googledrive::drive_deauth()
#'
#'   ## Two ways to get the Google Drive ID of the latest DoH Data Drop
#'   datadrop_id_latest()
#'   datadrop_id()
#'
#'   ## Two ways to get the Google Drive ID of the archive DoH Data Drop for
#'   ## 1 November 2020
#'   datadrop_id_archive(.date = "2020-11-01")
#'   datadrop_id(version = "archive", .date = "2020-11-01")
#' }
#'
#' @rdname datadrop_id
#' @export
#'
#
################################################################################

datadrop_id_latest <- function(verbose = TRUE) {
  ## Deauthorise
  #googledrive::drive_deauth()

  ## Get current data link folder information and contents
  dropCurrent <- googledrive::drive_ls(
    path = googledrive::drive_get(id = "1ZPPcVU4M7T-dtRyUceb0pMAd8ickYf8o")
  )

  ## Get dropDate
  dropDate <- get_drop_date(tbl = dropCurrent)

  if(verbose) {
    ## Provide message to user
    message(
      paste(
        strwrap(
          x = paste("Getting the Google Drive ID for the latest available DoH
                     Data Drop for ",
                    dropDate, ".",
                    sep = ""),
          width = 80,
        ),
        collapse = "\n"
      )
    )
  }

  ## Create temporary file
  destFile <- tempfile()

  datadrop_download(id = dropCurrent$id,
                    path = destFile,
                    overwrite = TRUE,
                    verbose = verbose)

  ## Get Google Drive ID for latest DoH Data Drop from readme file
  x <- get_id(.pdf = destFile)

  ## return x
  return(x)
}


################################################################################
#
#'
#' @rdname datadrop_id
#' @export
#'
#
################################################################################

datadrop_id_archive <- function(verbose = TRUE,
                                .date = NULL) {
  ## Deauthorise
  #googledrive::drive_deauth()

  ## Check if .date is not NULL
  if(is.null(.date)) {
    stop(
      paste(
        strwrap(
          x = "Date needs to be specified if version is set to archive. Please try again.",
          width = 80
        ),
        collapse = "\n"
      ),
      call. = TRUE
    )
  }

  ## Check if .date is current date
  if(lubridate::ymd(.date) == Sys.Date()) {
    stop(
      paste(
        strwrap(
          x = "Date should be earlier than current date to access data archive. Please try again.",
          width = 80
        ),
        collapse = "\n"
      ),
      call. = TRUE
    )
  }

  ## Check whether date is within range
  if(!lubridate::ymd(.date) %within% lubridate::interval(lubridate::ymd("2020-04-14"), Sys.Date())) {
    stop(
      paste(
        strwrap(
          x = "Earliest COVID-19 Data Drop record is for 2020-04-14. Only provide
                 dates as early as 2020-04-14 or later. Please try again.",
          width = 80
        ),
        collapse = "\n"
      ),
      call. = TRUE
    )
  }

  ## Provide message to user if verbose
  if(verbose) {
    message(
      paste(
        strwrap(
          x = paste("Getting the Google Drive ID for the DoH Data Drop archive for ",
                    .date, sep = ""),
          width = 80,
        ),
        collapse = "\n"
      )
    )
  }

  ## Get ID
  x <- "bit.ly/DataDropArchives" %>%
    RCurl::getURL() %>%
    stringr::str_extract_all(pattern = "[A-Za-z0-9@%#&()+*$,._\\-]{33}") %>%
    unlist()

  y <- googledrive::drive_get(id = x) %>%
    googledrive::drive_ls() %>%
    dplyr::filter(name == "COVID-19 DATA") %>%
    dplyr::select(id) %>%
    as.character()

  ## List contents of Data Drop Archive Google Drive Folder
  dropArchive <- googledrive::drive_get(id = y) %>%
    googledrive::drive_ls()

  ## Get gdriveID based on date
  dropArchiveDate <- dropArchive$name %>%
    lubridate::parse_date_time(orders = "my")

  dropArchiveDate <- paste(lubridate::month(dropArchiveDate),
                           lubridate::year(dropArchiveDate), sep = "/")

  ## Create specified date marker for .date
  month_year <- .date %>%
    lubridate::ymd() %>%
    lubridate::month() %>%
    paste(lubridate::year(lubridate::ymd(.date)), sep = "/")

  ## Process date
  collapsedDate <- stringr::str_remove_all(string = .date, pattern = "-")

  ## Check if archive contains folder for data for the specified month/year
  if(month_year %in% dropArchiveDate) {
    ## Get drive ID for folder of month_year
    month_year_id <- dropArchive$id[dropArchiveDate == month_year]

    ## Get month_year contents
    month_year_contents <- googledrive::drive_get(id = month_year_id) %>%
      googledrive::drive_ls()

    ## Check if collapsedDate is in month_year_contents
    if(any(stringr::str_detect(string = month_year_contents$name,
                               pattern = collapsedDate))) {
      archiveID <- month_year_contents %>%
        dplyr::filter(stringr::str_detect(string = name,
                                          pattern = collapsedDate)) %>%
        dplyr::select(id) %>%
        as.character()
    } else {
      warning(
        paste(
          strwrap(
            x = paste("Data Drop archives do not contain a folder for the day of ",
                      .date, ". Returning NULL.", sep = ""),
            width = 80
          ),
          collapse = "\n"
        ),
        call. = FALSE
      )

      ## Set archiveID to NULL
      archiveID <- NULL
    }
  } else {
    warning(
      paste(
        strwrap(
          x = paste("Data Drop archives do not contain a folder for the day of ",
                    .date, ". Returning NULL.", sep = ""),
          width = 80
        ),
        collapse = "\n"
      ),
      call. = FALSE
    )

    ## Set archiveID to NULL
    archiveID <- NULL
  }

  ## return archiveID
  return(archiveID)
}


################################################################################
#
#'
#' @rdname datadrop_id
#' @export
#'
#
################################################################################

datadrop_id <- function(verbose = TRUE,
                        version = c("latest", "archive"),
                        .date = NULL) {
  ## Get version
  version <- match.arg(arg = version)

  ## Get id
  if(version == "latest") {
    id <- datadrop_id_latest(verbose = verbose)
  } else {
    id <- datadrop_id_archive(verbose = verbose, .date = .date)
  }

  ## return id
  return(id)
}


################################################################################
#
#'
#' Get Google Drive ID for specified file in DoH Data Drop
#'
#' @param tbl A tibble output produced by [datadrop_ls()] that lists the files
#'   within a particular **DoH Data Drop** *Google Drive* folder
#' @param fn A character string composed of a word or words that can be used to
#'   match to the name of a file within a particular **DoH Data Drop**
#'   *Google Drive* folder listed in `tbl`.
#'
#' @return A 33-character string for the *Google Drive* ID of the specified
#'   **DoH Data Drop** file. If `fn` matches with more than one file, a vector of
#'   33-character strings for the Google Drive IDs of the specified DoH Data
#'   Drop files.
#'
#' @author Ernest Guevarra
#'
#' @examples
#' \dontrun{
#'   library(googledrive)
#'
#'   ## Authentication
#'   googledrive::drive_auth_configure(api_key = Sys.getenv("GOOGLEDRIVE_TOKEN"))
#'
#'   ## Deauthorise
#'   googledrive::drive_deauth()
#'
#'   ## Typical workflow
#'   tbl <- datadrop_ls(id = datadrop_id())
#'   datadrop_id_file(tbl = tbl, fn = "Case Information")
#'
#'   ## Piped workflow using magrittr %>%
#'   library(magrittr)
#'
#'   ## Get the id for the latest Case Information file
#'   datadrop_id() %>%
#'     datadrop_ls() %>%
#'     datadrop_id_file(fn = "Case Information")
#' }
#'
#' @rdname datadrop_id
#' @export
#'
#
################################################################################

datadrop_id_file <- function(tbl, fn) {
  ## Check if fn is found in tbl$name
  if(any(stringr::str_detect(string = tbl[["name"]], pattern = fn))) {
    ##
    id <- tbl %>%
      filter(stringr::str_detect(string = name, pattern = fn)) %>%
      select(id) %>%
      as.character()
  } else {
    ## Set id to NULL
    id <- NULL
    warning(
      paste(
        strwrap(
          x = paste("File/s with the word/s ", fn, " was not found in the Data
                    Drop folder. Please revise as needed and try again. Returning
                    NULL.", sep = ""),
          width = 80
        ),
        collapse = "\n"
      )
    )
  }

  ##
  return(id)
}
